enhancements/network/core-network-policies.md

```diff
@@ -0,0 +1,532 @@
+api-approvers:
+  - No new API
+tracking-link:
+  - "https://issues.redhat.com/browse/OCPSTRAT-819"
+creation-date: 2024-08-13
+last-updated: 2024-11-14
+---
+
+# Network Policies for OpenShift Components
+
+
+## Summary
+
+Kubernetes NetworkPolicy objects allow control over what traffic can
+enter and leave a pod.  Currently, most OpenShift operators do not
+install policies for their namespaces.  This is a vital tool for
+network security that we should use.
+
+## Motivation
+
+OpenShift should use the OpenShift platform capabilities to secure the
+OpenShift platform workloads (including layered products that are
+considered part of the platform).  It should both restrict what can
+talk to our pods, but also what our pods can talk to, so that a
+compromised pod can not be used as a springboard for other attacks.
+
+
+### User Stories
+
+- As an administrator, I want to have confidence that the cluster
+  components are as secure as possible, and are not likely to be an
+  attack vector so that I can trust the platform.
+
+- As an administrator, I need a way to be able to disable the
+  OpenShift policies in case something goes wrong so that I can
+  operate my cluster as it was before while the policies are being
+  fixed
+
+- As an administrator, I need to be able to override specific
+  OpenShift policies to be more restrictive so that I can satisfy my
+  security department
+
+- As an OpenShift release manager, I want to know that all cluster
+  namespaces have policies applied so that I can determine if we are
+  conforming with the policies outlined in this document.
+
+- As an OpenShift support engineer, I need to know if network policy
+  is breaking a service (perhaps not even the service that surfaces
+  the error) before I spend a lot of time debugging things so that I
+  can efficiently diagnose the issue.
+
+
+### Goals
+
+- Make it easy for operators to add network policy support
+
+- Document how to override OpenShift policies
+
+- Audit the traffic to ensure that a policy is not breaking a service
+  and report it in a must-gather
+
+- Audit the policies for compliance
+
+- Ensure all OpenShift namespaces have a default deny-all policy
+  defined, and identify those that do not (including “openshift-”
+  namespaces managed by a non Red Hat operator)
+
+
+### Non-Goals
+
+- We are not considering requiring the definition of Admin Network
+  Policy for components (yet).  But the documentation of the network
+  connections that a namespace uses should make future development of
+  Admin Network Policy easier.  We will be using Admin Network Policy
+  to allow cluster admins to override our namespace Network Policies
+  if needed.
+
+- Adding policies for operators not created by Red Hat
+
+
+## Proposal
+
+1. Change the namespace admission controller so all namespaces with an
+   openshift- prefix (since those are special anyway) are labeled with
+   the “openshift-namespace’ label so **that network policy can address
+   them**
```

Comment:
The rule is horrible.

They get installed into a default openshift-operators namespace, but the OLM will need to own the policy for that namespace.  I suspect it will need to be wildcard allow egress and ingress.

But, worse, the admin can chose any namespace to add it to...  there are ongoing conversations about how to deal with this.

Reply:
Former OLM team member here (recently moved to control plane group) :wave: 

OLM installed operators probably won't be able to specify `NetworkPolicy` resources in their bundles according to https://olm.operatorframework.io/docs/tasks/creating-operator-manifests/#packaging-additional-objects-alongside-an-operator 

OLMv1 is supposed to GA in 4.18 and will, initially, respect a subset of the existing bundle format. The desire is to _eventually_ support bundling formats that allows operator authors more control as to the contents packaged within their bundle (OLMv1 is going to be unopinionated in _what_ is in the bundles) meaning operator authors will eventually be able to specify `NetworkPolicy` resources in their bundles. 

With the current bundling format it seems like OLM would need to be updated to either:
1. Allow `NetworkPolicy` resources in the bundle alongside the CSV. IIUC, to achieve compliance with the high-level goal of this proposal this would likely mean updating all bundles in the catalogs to contain a sufficient `NetworkPolicy` resource. Some open questions for this approach:
    - Do we have enough control of the catalog curation pipelines to ensure this is done properly?
    - Should a wildcard `NetworkPolicy` be added to all bundles by default if the authors of the operator can't/won't add one themselves?
2. Stamp out a permissive `NetworkPolicy` resource on each operator installation.

With the focus on OLMv1 and some general attrition, there are not many folks on the OLM team left with knowledge of the legacy "OLMv0". Both of these are likely not trivial asks of the OLM team.

I do wonder if having CVO manage a default `NetworkPolicy` for the `openshift-operators` namespace, that is permissive enough for all operators installed  in that namespace, is sufficient enough to achieve the compliance we are looking for even if admins can install an operator in any namespace?